pkg/envoy/resource_monitor_linux.go

```diff
@@ -0,0 +1,767 @@
+	if err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			return errors.New("memory controller not enabled")
+		}
+		return fmt.Errorf("failed to stat cgroup: %w", err)
+	}
+	if !info.IsDir() {
+		return fmt.Errorf("%s is not a directory", memoryPath)
+	}
+	return nil
+}
+
+var _ CgroupDriver = (*cgroupV1Driver)(nil)
+
+func DetectCgroupDriver() (CgroupDriver, error) {
+	const cgv2Magic = 0x63677270
+
+	fsType := func(path string) (int64, error) {
+		for {
+			var stat syscall.Statfs_t
+			err := syscall.Statfs(path, &stat)
+			if err != nil {
+				if errors.Is(err, syscall.EINTR) {
+					continue
+				}
+				return 0, err
+			}
+			return stat.Type, nil
+		}
+	}
+	osFs := os.DirFS("/")
+
+	// fast path: cgroup2 only
+	t, err := fsType("/sys/fs/cgroup")
+	if err != nil {
+		return nil, err
+	}
+	if t == cgv2Magic {
+		return &cgroupV2Driver{root: "sys/fs/cgroup", fs: osFs}, nil
+	}
+
+	// find the hybrid mountpoint, or fall back to v1
+	mountpoint, isV2, err := findMountpoint(osFs)
+	if err != nil {
+		return nil, err
+	}
+	if isV2 {
+		return &cgroupV2Driver{root: mountpoint, fs: osFs}, nil
+	}
+	return &cgroupV1Driver{root: mountpoint, fs: osFs}, nil
+}
+
+func parseCgroupName(contents []byte) (string, error) {
+	scan := bufio.NewScanner(bytes.NewReader(contents))
+	for scan.Scan() {
+		line := scan.Text()
+		if strings.HasPrefix(line, "0::") {
+			return strings.Split(strings.TrimPrefix(strings.TrimSpace(line), "0::"), " ")[0], nil
+		}
+	}
+	return "", errors.New("cgroup not found")
+}
+
+func findMountpoint(fsys fs.FS) (mountpoint string, isV2 bool, err error) {
+	mounts, err := fs.ReadFile(fsys, fmt.Sprintf("proc/%d/mountinfo", os.Getpid()))
+	if err != nil {
+		return "", false, err
+	}
+	scanner := bufio.NewScanner(bytes.NewReader(mounts))
+	var cgv1Root string
+	for scanner.Scan() {
+		line := strings.Fields(scanner.Text())
+		fsType := line[slices.Index(line, "-")+1]
+		switch fsType {
+		case "cgroup2":
+			return line[4][1:], true, nil
+		case "cgroup":
+			if cgv1Root == "" {
+				cgv1Root = filepath.Dir(line[4][1:])
+			}
+		}
+	}
+	if cgv1Root == "" {
+		return "", false, errors.New("no cgroup mount found")
+	}
+	return cgv1Root, false, nil
+}
+
+func marshalAny(msg proto.Message) *anypb.Any {
+	data := new(anypb.Any)
+	_ = anypb.MarshalFrom(data, msg, proto.MarshalOptions{
+		AllowPartial:  true,
+		Deterministic: true,
+	})
+	return data
+}
+
+type memoryLimitWatcher struct {
+	limitFilePath string
+
+	value atomic.Uint64
+
+	watches sync.WaitGroup
+}
+
+func (w *memoryLimitWatcher) Value() uint64 {
+	return w.value.Load()
+}
+
+func (w *memoryLimitWatcher) readValue() (uint64, error) {
+	data, err := os.ReadFile(w.limitFilePath)
+	if err != nil {
+		return 0, err
+	}
+	max := strings.TrimSpace(string(data))
+	if max == "max" {
+		return 0, nil
+	}
+	return strconv.ParseUint(max, 10, 64)
+}
+
+func (w *memoryLimitWatcher) Watch(ctx context.Context) error {
+	fd, err := syscall.InotifyInit1(syscall.IN_CLOEXEC)
```

Comment:
We have a file watcher in `fileutil`. But I guess its made for more of the cross-platform case and would also start a polling watcher, which is not ideal. Would using the [underlying package](namespacelabs.dev/go-filenotify) simplify things?  Or is it really no different?

Reply:
Sure, I did this manually since the usage here is simple enough maybe not to warrant using a library for, but I can rewrite to use github.com/fsnotify/fsnotify if you'd like. It might make the code a bit more readable.

Reply:
Sure, I did this manually since the usage here is simple enough maybe not to warrant using a library for, but I can rewrite to use github.com/fsnotify/fsnotify if you'd like. It might make the code a bit more readable.

Reply:
No preference from me. Whichever way seems best.

Reply:
Looks like there are some wrappers around the low level syscall functions in golang.org/x/sys/unix that I can use here, that might be good for now.